app/services/sp_return_url_resolver.rb

```diff
@@ -0,0 +1,45 @@
+class SpReturnUrlResolver
+  attr_reader :sp, :sp_request_url, :view_context
+
+  def initialize(sp:, sp_request_url:)
+    @sp = sp
+    @sp_request_url = sp_request_url
+  end
+
+  def return_to_sp_url
+    if oidc_redirect_uri_present?
+      oidc_access_denied_redirect_url
+    elsif sp.return_to_sp_url.present?
+      sp.return_to_sp_url
+    else
+      inferred_redirect_url
+    end
+  end
+
+  def failure_to_proof_url
+    sp.failure_to_proof_url.presence || return_to_sp_url
+  end
+
+  private
+
+  def inferred_redirect_url
+    configured_url = sp.redirect_uris&.first || sp.acs_url
+    URI.join(configured_url, '/').to_s
+  end
+
+  def oidc_access_denied_redirect_url
+    UriService.add_params(
+      sp_request_url_pararms[:redirect_uri],
+      error: 'access_denied',
+      state: sp_request_url_pararms[:state],
```

Comment:
could we simplify this class if we made the params `state` and `redirect_uri` instead of passing in `sp_request_uri` and parsing it?

Reply:
Also should we be checking that `sp.redirect_uris.include?(redirect_uri_param)` here too, to validate it?

Reply:
I don't think we need to validate it here. It's validated at the authorize endpoint when we save all of this.

Reply:
That said, we can add an additional layer of validation. I don't think it will hurt. WDYT?

Reply:
> I don't think we need to validate it here. It's validated at the authorize endpoint when we save all of this.

> That said, we can add an additional layer of validation. I don't think it will hurt. WDYT?

I don't feel that strongly, but you're right, since it's validated in the other spot we're covered

Reply:
I marked this resolved, but made an interested discovery reworking the tests. If you provide a redirect uri and state param in a SAML request they won't be validated, but they will be respected by this code. Those URLs would be used as the cancellation links. That is for sure a cause for concern which I'll add some kind of validation to address.

Reply:
Addressed. Here's the test the helped me catch this: https://github.com/18F/identity-idp/blob/9220ea781cc6c99e35d4f955b78d42eef4d47751/spec/features/saml/redirect_uri_validation_spec.rb#L7